XXX-asyncwrap-api.md

```diff
@@ -503,52 +470,76 @@ This cannot be enforced in any way. It is up to the implementer to make sure
 all of the callbacks are placed and called at the correct time.
 
 
-#### `async_wrap.createEvent(handle[, parent_id])`
-
-* `handle` {Object}
-* `parent_id` {Number} **Default:** `async_wrap.currentId()`
-* Return: {AsyncEvent}
+#### `async_hooks.newUid()`
 
-Returns an instance of `AsyncEvent` to be used to trigger `AsyncHook` callback.
+* Return: {Number}
 
-`handle` is the object instance that's passed to `init()`.
+Return a new unique id for a given handle from the same pool of unique id's
+used for all internally created handles.
 
-`parent_id` is the value passed as `parentId` to `init()`. If omitted then the
-current async id is used.
+Generally this should be used during object construction. e.g.:
 
-An instance of `AsyncEvent` is returned that is to be used to trigger the
-`AsyncHook` callbacks.
+```js
+class MyClass {
+  constructor() {
+    this._event_id = async_hooks.newUid();
+  }
+}
+```
 
 
-#### `event.getId()`
+#### `async_hooks.emitInit(id, handle, type[, parentId[, parentEvent]])`
 
-* Return: {Number}
-
-Returns the unique id assigned to the instance when it was created.
+* `id` {Number} Generated by calling `newUid()`
+* `handle` {Object}
+* `type` {String}
+* `parentId` {Number} **Default:** `currentId()`
+* `parentEvent` {Object} **Default:** `undefined`
+* Return: {Object|Null}
 
-Would be easier to just slap on the number, but right now I'm not 100% sure
-that the implementation won't change in the future. So sticking with this.
+Emit that a handle is being initialized. `id` should be a value returned by
+`async_hooks.newUid()`. Usage will probably be as follows:
 
+```js
+class Foo {
+  constructor() {
+    this.event_id = async_hooks.newUid();
+    this.event_stor = async_hooks.emitInit(this.event_id, this, 'Foo');
+  }
+}
+```
 
-#### `event.emitInit()`
+The return value from `emitInit()` is either an instance of the internal
+constructor `AsyncEventStor`. Which will contain the state of all hooks that
+must propagate to future async calls. If there are no hooks that need to
+propagate then the return value will be `null`. Preventing the need of creating
+one additional object for every async constructor, even when hooks aren't being
+used.
 
-* Return: {Undefined}
+In the unusual circumstance that the embedder needs to define a different
+parent id than `currentId()`, they can pass in that id manually. Along with
+the `AsyncEventStor`, if there is one. If no stor is passed then it is assumed
+there are no hooks that need to propagate to the newly created stor.
 
-Trigger listening `init()` callbacks that this event has been created.  Will
-throw if `emitInit()` is called twice.
+It is suggested to have `emitInit()` be the last call in the object's
+constructor.
 
 
-#### `event.emitBefore()`
+#### `async_hooks.emitBefore(id, event)`
```

Comment:
what happens if you:

- `emit(<any positive number>, null)`
- `emit(<any positive number>, {})`
- `emit(-1, <anything>)`
- `emit(0, <anything>)`